File: backend/app/schemas/company_billing.py
"""Schemas for the company-facing billing portal API (app/routers/company_billing.py)."""
from datetime import datetime

from pydantic import BaseModel, EmailStr, Field, field_validator


# ---------------- Profile ----------------

class CompanyBillingProfileOut(BaseModel):
    name: str
    gstin: str
    legal_name: str
    billing_address: str
    state_code: str


class CompanyBillingProfileUpdate(BaseModel):
    # Lengths are bounded here so an oversized value 422s at the edge.
    # gstin/legal_name match their columns (companies.gstin String(20),
    # legal_name String(255)) — over-length is a 500 on Postgres and a
    # silently stored overflow on SQLite. billing_address is a Text column,
    # so its cap is a policy limit that keeps an unbounded blob out of the
    # buyer block of every rendered tax invoice.
    gstin: str | None = Field(default=None, max_length=20)
    legal_name: str | None = Field(default=None, max_length=255)
    billing_address: str | None = Field(default=None, max_length=2000)
    state_code: str | None = None

    @field_validator("state_code")
    @classmethod
    def _validate_state_code(cls, v):
        if v is None:
            return v
        if v == "" or (len(v) == 2 and v.isdigit()):
            return v
        raise ValueError("state_code must be a 2-digit code (e.g. '29') or empty")


# ---------------- Invoices ----------------

class CompanyInvoiceItemOut(BaseModel):
    description: str
    quantity: int
    unit_price: float
    line_total: float


class CompanyInvoiceOut(BaseModel):
    id: int
    invoice_number: str | None = None
    status: str
    subtotal: float
    cgst: float
    sgst: float
    igst: float
    total: float
    tax_note: str
    due_date: datetime | None = None
    issued_at: datetime | None = None
    paid_at: datetime | None = None
    notes: str
    items: list[CompanyInvoiceItemOut]


# ---------------- Seat pools ----------------

class SeatPoolOut(BaseModel):
    id: int
    course_id: int | None = None
    bundle_id: int | None = None
    course_title: str | None = None
    bundle_name: str | None = None
    total_seats: int
    used_seats: int


class SeatAssignRequest(BaseModel):
    email: EmailStr


class SeatAssignmentOut(BaseModel):
    user_id: int
    email: str
    assigned_at: datetime | None = None
    granted_course_ids: list[int] = Field(default_factory=list)
    already_had_course_ids: list[int] = Field(default_factory=list)

File: frontend/src/api/companyBilling.ts
import { api } from './axios'

// ============================================================================
// Types
// ============================================================================

export interface BillingProfile {
  name: string
  gstin: string
  legal_name: string
  billing_address: string
  state_code: string
}

export interface BillingProfileUpdate {
  gstin?: string
  legal_name?: string
  billing_address?: string
  state_code?: string
}

export type InvoiceStatus = 'issued' | 'paid' | 'cancelled'

export interface InvoiceItem {
  description: string
  quantity: number
  unit_price: number
  line_total: number
}

export interface Invoice {
  id: number
  invoice_number: string | null
  status: InvoiceStatus
  subtotal: number
  cgst: number
  sgst: number
  igst: number
  total: number
  tax_note: string
  due_date: string | null
  issued_at: string | null
  paid_at: string | null
  notes: string
  items: InvoiceItem[]
}

export interface InvoicePayOrder {
  order_id: string
  amount: number
  currency: string
  key_id: string
}

export interface SeatPool {
  id: number
  course_id: number | null
  bundle_id: number | null
  course_title: string | null
  bundle_name: string | null
  total_seats: number
  used_seats: number
}

export interface SeatAssignment {
  user_id: number
  email: string
  assigned_at: string | null
  granted_course_ids?: number[]
  already_had_course_ids?: number[]
}

// ============================================================================
// Profile
// ============================================================================

export const fetchBillingProfile = async (): Promise<BillingProfile> => {
  const { data } = await api.get('/companies/billing/profile')
  return data
}

export const updateBillingProfile = async (
  update: BillingProfileUpdate
): Promise<BillingProfile> => {
  const { data } = await api.patch('/companies/billing/profile', update)
  return data
}

// ============================================================================
// Invoices
// ============================================================================

export const fetchInvoices = async (): Promise<Invoice[]> => {
  const { data } = await api.get('/companies/billing/invoices')
  return data
}

/**
 * Download an invoice PDF via an authenticated GET (the file lives behind
 * auth, so a plain <a href> would 401 — fetch as a blob, wrap it in an
 * object URL, and click a throwaway anchor).
 */
export const downloadInvoicePdf = async (invoiceId: number, filename?: string): Promise<void> => {
  const { data } = await api.get(`/companies/billing/invoices/${invoiceId}/pdf`, {
    responseType: 'blob',
  })
  const blob = new Blob([data], { type: 'application/pdf' })
  const url = window.URL.createObjectURL(blob)
  const link = document.createElement('a')
  link.href = url
  link.download = filename || `invoice-${invoiceId}.pdf`
  document.body.appendChild(link)
  link.click()
  document.body.removeChild(link)
  window.URL.revokeObjectURL(url)
}

export const createInvoiceOrder = async (invoiceId: number): Promise<InvoicePayOrder> => {
  const { data } = await api.post(`/companies/billing/invoices/${invoiceId}/pay`)
  return data
}

export const verifyInvoicePayment = async (p: {
  razorpay_order_id: string
  razorpay_payment_id: string
  razorpay_signature: string
  invoice_id: number
}): Promise<{ success: boolean; message: string }> => {
  const { data } = await api.post('/payments/verify', p)
  return data
}

// ============================================================================
// Seat pools
// ============================================================================

export const fetchSeatPools = async (): Promise<SeatPool[]> => {
  const { data } = await api.get('/companies/billing/seat-pools')
  return data
}

export const assignSeat = async (poolId: number, email: string): Promise<SeatAssignment> => {
  const { data } = await api.post(`/companies/billing/seat-pools/${poolId}/assign`, { email })
  return data
}

export const fetchSeatAssignments = async (poolId: number): Promise<SeatAssignment[]> => {
  const { data } = await api.get(`/companies/billing/seat-pools/${poolId}/assignments`)
  return data
}
